src/proposer/proposer_rpc.cpp

```diff
@@ -167,6 +170,126 @@ class ProposerRPCImpl : public ProposerRPC {
     obj.pushKV("stakeable_coins", arr);
     return obj;
   }
+
+  UniValue ProposeBlocks(CWallet &wallet,
+                         const boost::optional<CScript> &coinbase_script,
+                         const int num_generate) const {
+
+    proposer::Proposer *proposer = GetComponent<proposer::Proposer>();
+
+    UniValue block_hashes(UniValue::VARR);
+
+    // To pick up to date coins for staking we need to make sure that the wallet is synced to the current chain.
+    wallet.BlockUntilSyncedToCurrentChain();
+
+    esperanza::WalletExtension &wallet_ext = wallet.GetWalletExtension();
+
+    for (int i = 0; i < num_generate; ++i) {
+
+      std::shared_ptr<const CBlock> block;
+      {
```

Comment:
The scope is redundant.

Reply:
Why? I cannot keep the lock on `cs_main` while doing `ProcessNewBlock()`

Reply:
Yep, sorry. I misread the scope borders in GitHub UI.